backend/test/test_czi_hosted/unit/common/utils/test_cache.py

```diff
@@ -0,0 +1,248 @@
+
+    def test_write_lock_demoted_to_read_lock_when_data_acquired(self):
+        cache_item = CacheItem()
+        some_data = {"some": "data"}
+        cache_item.get("key", lambda key: some_data)
+        self.assertFalse(cache_item.data_lock.d_lock.locked())
+        self.assertEqual(cache_item.data_lock.num_r, 1)
+        self.assertTrue(cache_item.data_lock.w_lock.locked())
+
+    def test_write_lock_demoted_to_read_when_exception_thrown_retrieving_data(self):
+        def raise_exception(cache_key):
+            raise AssertionError
+
+        cache_item = CacheItem()
+        with self.assertRaises(AssertionError):
+            cache_item.get("key", raise_exception)
+        self.assertFalse(cache_item.data_lock.d_lock.locked())
+        self.assertEqual(cache_item.data_lock.num_r, 1)
+        self.assertTrue(cache_item.data_lock.w_lock.locked())
+
+    def test_data_not_deleted_when_lock_held(self):
+        cache_item = CacheItem()
+        some_data = {"some": "data"}
+        cache_item.get("key", lambda key: some_data)
+        # check the lock
+        self.assertEqual(cache_item.data_lock.num_r, 1)
+        deleted = cache_item.attempt_delete()
+        self.assertFalse(deleted)
+        # check data still available
+        data = cache_item.get("key")
+        self.assertEqual(data, some_data)
+
+        # drop locks and try deleting
+        cache_item.release()
+        cache_item.release()
+
+        deleted = cache_item.attempt_delete()
+        self.assertTrue(deleted)
+
+        # check data deleted
+        data = cache_item.get("key")
+        self.assertIsNone(data)
+
+    def test_deletion_idempotent(self):
+        cache_item = CacheItem()
+        cache_item.get("cache_key", self.create_data)
+        cache_item.attempt_delete()
+        # this should not throw an error
+        cache_item.attempt_delete()
+
+class CacheManagerTest(unittest.TestCase):
+
+    def test_cache_manager_retrieves_cached_data(self):
+        cache_manager = CacheManager(max_cached=3, timelimit_s=2)
+        get_data = MagicMock()
+        get_data.return_value = {"cache_key": "cache_key"}
+        with cache_manager.get("key_one", lambda x: {x: x}) as cache_item:
+            self.assertIsNotNone(cache_item)
+        with cache_manager.get("key_one", get_data) as cache_item:
+            self.assertIsNotNone(cache_item)
+            self.assertEqual(get_data.call_count, 0)
+
+    def test_cache_manager_deletes_old_datasets(self):
+        cache_manager = CacheManager(max_cached=3, timelimit_s=2)
+        get_data = MagicMock()
+        get_data.return_value = {"cache_key": "cache_key"}
+        with cache_manager.get("key_one", lambda x: {x: x}) as cache_item:
+            self.assertIsNotNone(cache_item)
+        with cache_manager.get("key_one", get_data) as cache_item:
+            self.assertIsNotNone(cache_item)
+            self.assertEqual(get_data.call_count, 0)
+        sleep(3)
```

Comment:
the '3' is a bit mysterious.  I suggest:
* define a constant called 'cacheTimeLimit_s' and set to 2.
* use that variable in the CacheManager constructor
* use that variable in sleep, eg, `sleep(cacheTimeLimit_s + 1)`
Makes it more obvious what you are doing, and makes it easier to keep the code in sync